File: fsharp/AoC2021/Day05/Day05/Solution.fs
module Solution

type Point =
    { X: int; Y: int }
    static member parse (str: string) =
        let idx = str.IndexOf(',')
        if idx = -1 then failwith $"Invalid input: {str}"
        { X = int str[..idx - 1]; Y = int str[(idx + 1)..] }

let asSeq x y =
    if x < y then
        { x .. y }
    elif x > y then
        { x .. -1 .. y }
    else
        Seq.initInfinite (fun _ -> x)

type Line =
    { P1: Point; P2: Point }
    static member parse (str: string) =
        let idx = str.IndexOf(" -> ")
        if idx = -1 then failwith $"Invalid input: {str}"
        { P1 = Point.parse str[..idx - 1]; P2 = Point.parse str[(idx + 4)..] }
    member this.isStraight =
        this.P1.X = this.P2.X || this.P1.Y = this.P2.Y
    member this.points =
        (asSeq this.P1.X this.P2.X, asSeq this.P1.Y this.P2.Y ) ||> Seq.map2 (fun x y -> { X = x; Y = y })

let countDangerousPoints (includeDiagonals: bool) (lines: seq<Line>) =
    lines
    |> Seq.filter (fun line -> includeDiagonals || line.isStraight)
    |> Seq.collect (fun line -> line.points)
    |> Seq.groupBy id
    |> Seq.map (fun (_, points) -> points |> Seq.length)
    |> Seq.filter (fun x -> x > 1)
    |> Seq.length

File: csharp/AoC2021/Day05/Day05/Solution.cs
namespace Day05;

public static class Solution
{
    public static void Solve(IEnumerable<string> input)
    {
        var lines = input.Select(Line.Parse).ToList();

        var count1 = CountDangerousPoints(lines, false);
        Console.WriteLine($"Not considering diagonals, there are {count1} points where multiple lines overlap.");

        var count2 = CountDangerousPoints(lines, true);
        Console.WriteLine($"Considering diagonals, there are {count2} points where multiple lines overlap.");
    }

    public static int CountDangerousPoints(IEnumerable<Line> lines, bool includeDiagonals)
    {
        Dictionary<Point, int> counts = new();
        lines
            .Where(line => includeDiagonals || line.IsHorizontal() || line.IsVertical())
            .SelectMany(line => line.GetPoints())
            .ToList()
            .ForEach(point => counts[point] = counts.ContainsKey(point) ? counts[point] + 1 : 1);
        return counts.Values.Count(c => c > 1);
    }
}
